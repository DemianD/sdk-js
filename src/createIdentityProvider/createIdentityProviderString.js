function encodeParams(options = {}) {
  return Object.keys(options)
    .map(key => `${key}=${encodeURIComponent(options[key])}`)
    .join('&')
}

export default function createIdentityProviderString(type, options) {
  return `${type}://${encodeParams(options)}`
}
