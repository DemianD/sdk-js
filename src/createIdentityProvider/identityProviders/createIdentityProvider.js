import retry from '../../util/retry'

import { createIdentityProviderError } from '../../util/createError'

import {
  MISCONFIGURED,
  MISCONFIGURED_REALM,
  MISCONFIGURED_AUTH_PROVIDER_OPTIONS,
} from '../../constants'
import { joinAsSpeech, AND } from '../../util/joinAsSpeech'
import omit from '../../util/omit'

const MAX_ATTEMPTS = 3
const REFETCH_WINDOW = 20

function secondsToMilliseconds(seconds) {
  return seconds * 1000
}

export function createIdentityProvider(
  grant_type,
  requiredParameters = [],
  optionalParameters = []
) {
  return class {
    constructor(network, parameters = {}) {
      this._network = network
      this._parameters = parameters

      // Check if we got a realm
      const { realm } = parameters
      if (!realm) {
        throw createIdentityProviderError(
          `realm is a required option for \`${grant_type}\``,
          MISCONFIGURED,
          MISCONFIGURED_REALM
        )
      }

      const allPossibleParameters = [
        ...requiredParameters,
        ...optionalParameters,
      ]

      const givenParameters = Object.keys(parameters).filter(
        key => key !== 'realm' // Skip the realm option, will be validated by its own
      )

      const missingRequiredParameters = requiredParameters.filter(
        required => !givenParameters.includes(required)
      )

      const extraMessage = `\n  - Required parameters: ${joinAsSpeech(
        AND,
        requiredParameters
      ) || '<none>'}\n  - Optional parameters: ${joinAsSpeech(
        AND,
        optionalParameters
      ) || '<none>'}`

      if (missingRequiredParameters.length > 0) {
        throw createIdentityProviderError(
          `${joinAsSpeech(AND, missingRequiredParameters)} ${
            missingRequiredParameters.length === 1
              ? 'is a required parameter'
              : 'are required parameters'
          } for \`${grant_type}\`${extraMessage}`,
          MISCONFIGURED,
          MISCONFIGURED_AUTH_PROVIDER_OPTIONS
        )
      }

      const unnecessaryParameters = givenParameters.filter(
        parameter => !allPossibleParameters.includes(parameter)
      )

      if (unnecessaryParameters.length > 0) {
        throw createIdentityProviderError(
          `${joinAsSpeech(AND, unnecessaryParameters)} ${
            unnecessaryParameters.length === 1
              ? 'is not an optional or a required parameter'
              : 'are not optional or required parameters'
          } for \`${grant_type}\`${extraMessage}`,
          MISCONFIGURED,
          MISCONFIGURED_AUTH_PROVIDER_OPTIONS
        )
      }
    }

    getAuthorization() {
      const { realm } = this._parameters

      if (this._current === undefined) {
        this._current = retry(
          (resolve, reject) => {
            // Get the access_token
            this._network
              .post(
                `${realm}/access_tokens`,
                Object.assign({}, omit(this._parameters, ['realm']), {
                  grant_type,
                })
              )
              .then(resolve, reject)
          },
          {
            max_attempts: MAX_ATTEMPTS,
          }
        )
          // Setup all the required parts to authenticate
          .then(({ data }) => ({
            Authorization: `${data.token_type} ${data.access_token}`,
            Expiration: data.expires_in,
            Realm: realm,
          }))

          // Get the proxy url
          .then(({ Realm, Authorization, Expiration }) =>
            this._network
              .get(`${Realm}/integrations/proxy`, {
                headers: { Authorization },
              })
              .then(({ data: response }) => ({
                Realm: response.data.url,
                Authorization,
                Expiration,
              }))
          )

          // Make sure to create a new access_token when the current one is going to expire
          .then(access => {
            setTimeout(() => {
              this._current = undefined
              this.getAuthorization()
            }, secondsToMilliseconds(access.Expiration - REFETCH_WINDOW))
            return access
          })
      }

      return this._current
    }
  }
}
