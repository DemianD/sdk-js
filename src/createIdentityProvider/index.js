import createError from '../util/createError'
import { MISCONFIGURED, MISCONFIGURED_AUTH_PROVIDER } from '../constants'
import joinAsSpeech, { AND } from '../util/joinAsSpeech'

import Client from './Client'
import parse from './parseIdentityProviderString'

const IDPS = {
  client: Client,
}

export default function createIdentityProvider(idp_descriptor, idps = IDPS) {
  const { type, options } = parse(idp_descriptor)

  if (idps[type] === undefined) {
    const providers = Object.keys(idps).map(key => `\`${key}\``)

    throw createError(
      `Identity provider \`${type}\` does not exist. The only ${
        providers.length === 1 ? `option is` : `options are`
      } ${joinAsSpeech(AND, providers)}.`,
      MISCONFIGURED,
      MISCONFIGURED_AUTH_PROVIDER
    )
  }

  return new idps[type](options)
}
