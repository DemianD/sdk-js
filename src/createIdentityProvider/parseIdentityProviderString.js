import createError from '../util/createError'

export default function parseIdentityProviderString(idp_descriptor = '') {
  const [type, encoded_options] = idp_descriptor.split('://')

  if (type === undefined || type === '') {
    throw createError('You should provide a `type` as identity provider.')
  }

  if (encoded_options === undefined || encoded_options === '') {
    throw createError('You should provide `options` for the identity provider.')
  }

  const options = encoded_options
    .split('&')
    .map(item => item.split('='))
    .map(([key, value]) => ({
      [key]: decodeURIComponent(value),
    }))
    .reduce((a, b) => Object.assign(a, b))

  return { type, options }
}
