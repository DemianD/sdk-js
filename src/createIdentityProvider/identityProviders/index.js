import { createIdentityProvider } from './createIdentityProvider'
import Token from './Token'

export const IDPS = {
  client_credentials: createIdentityProvider(
    'client_credentials',
    ['client_id', 'client_secret'],
    ['scope']
  ),
  office365_code: createIdentityProvider('office365_code', [
    'code',
    'client_id',
  ]),
  password: createIdentityProvider(
    'password',
    ['client_id', 'client_secret', 'username', 'password'],
    ['scope', 'enterprise_id']
  ),
  public_client: createIdentityProvider(
    'public_client',
    ['client_id'],
    ['scope']
  ),
  public_client_password: createIdentityProvider(
    'public_client_password',
    ['client_id', 'username', 'password'],
    ['scope', 'enterprise_id']
  ),
  public_client_password_reset: createIdentityProvider(
    'public_client_password_reset',
    ['client_id', 'username'],
    ['password']
  ),
  resource_code: createIdentityProvider('resource_code', ['client_id', 'code']),
  saml: createIdentityProvider('saml', [
    'payload',
    'enterprise_id',
    'client_id',
  ]),

  token: Token,
}
