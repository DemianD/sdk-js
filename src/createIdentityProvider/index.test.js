import createIdentityProvider from '.'

describe('createIdentityProvider', () => {
  it('should error when a the idp descriptor is not defined or invalid', () => {
    expect(() => createIdentityProvider()).toThrowErrorMatchingSnapshot()
    expect(() => createIdentityProvider('')).toThrowErrorMatchingSnapshot()
    expect(() =>
      createIdentityProvider('example')
    ).toThrowErrorMatchingSnapshot()
  })

  it('should error when a provider does not exist', () => {
    expect(() =>
      createIdentityProvider('example://some=options')
    ).toThrowErrorMatchingSnapshot()

    expect(() =>
      createIdentityProvider('example://some=options', {
        client: 'client provider',
        direct: 'direct provider',
      })
    ).toThrowErrorMatchingSnapshot()
  })
})
