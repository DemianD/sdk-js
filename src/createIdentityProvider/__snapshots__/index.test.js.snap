// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createIdentityProvider should error when a provider does not exist 1`] = `"Identity provider \`example\` does not exist. The only option is \`client\`."`;

exports[`createIdentityProvider should error when a provider does not exist 2`] = `"Identity provider \`example\` does not exist. The only options are \`client\` and \`direct\`."`;

exports[`createIdentityProvider should error when a the idp descriptor is not defined or invalid 1`] = `"You should provide a \`type\` as identity provider."`;

exports[`createIdentityProvider should error when a the idp descriptor is not defined or invalid 2`] = `"You should provide a \`type\` as identity provider."`;

exports[`createIdentityProvider should error when a the idp descriptor is not defined or invalid 3`] = `"You should provide \`options\` for the identity provider."`;
